src/shogun/base/subset_iterators.h

```diff
@@ -0,0 +1,86 @@
+/*
+ * This software is distributed under BSD 3-clause license (see LICENSE file).
+ *
+ * Authors: Gil Hoben
+ */
+
+#ifndef SHOGUN_SUBSET_ITERATORS_H
+#define SHOGUN_SUBSET_ITERATORS_H
+
+#include <shogun/labels/Labels.h>
+#include <shogun/features/Features.h>
+
+namespace shogun
+{
+	template <
+	    typename T, std::enable_if<
+	                    (std::is_base_of<T, CFeatures>::value) ||
+	                    (std::is_base_of<T, CLabels>::value)>* = nullptr>
+	class SubsetIterator
+	{
+	public:
+		SubsetIterator(T* pointer, index_t idx = 0)
```

Comment:
Could the pointer be const?

Reply:
I was just thinking that instead we should pass a pointer to the matrix/vector and the subset pointer. that would make the iterator more conventional. I don't like that it holds a reference to the class...

Reply:
and that we to pass `this`

Reply:
No, some features don’t have matrices but a get feature vector method

Reply:
@karlnapf i removed the reference counting because it doesn't make sense for the iterator to have any ownership over the class. This should be a weak reference. If the feature object is deleted the user should not be using the iterator anymore anyway

Reply:
exactly! I cannot imagine a case where this would not be the case.
But the const would also be good for potential speed no?

Reply:
well we return by value so i am not sure it needs a const? ideally we would return a const ref..

Reply:
True It’s a new vector.
Just we want the make the vector access methods const later on and we want to avoid that people write into the vectors they retrieve

Reply:
ah yes true!

Reply:
btw back to the original question of passing const pointer, I don't think it is worth it. const pointer class members don't bring any (or at least negligible)  performance gains

Reply:
can this iterator be used when I only have const a pointer say `const CDenseFeatures<...>*`?

Reply:
but if it is const you wouldn't have subset? In that case need to add a const_iterator, should be pretty much the same

Reply:
just need to make get_subset_stack() const and then it should work!